### PR TITLE
short circuit case rebuilds

### PR DESCRIPTION
**Pull request by snopoke**
##### SUMMARY
Prevent full case rebuilds if there is no need to rebuild the case:
* there is no case transaction for the form
* the case did not already exist

I expect that 99% of the current rebuilds on ICDS will trigger the first short circuit (no case transactions for the form)

### CODE REVIEW

**Review by calellowitz — APPROVED** [review]

File: corehq/form_processor/backends/sql/processor.py [hunk 1]
```diff
@@ -210,56 +210,75 @@ def submission_error_form_instance(cls, domain, instance, message):
     def get_cases_from_forms(case_db, xforms):
         """Get all cases affected by the forms. Includes new cases, updated cases.
         """
-        touched_cases = {}
-        if len(xforms) > 1:
-            domain = xforms[0].domain
+        if len(xforms) == 1:
```
Comment (calellowitz) [inline corehq/form_processor/backends/sql/processor.py]: what happens if len(xforms) is 0. this appears to change that behavior

Comment (snopoke) [reply corehq/form_processor/backends/sql/processor.py]: that would previously have errored when running `xforms[0]`

**Comment by snopoke** [comment]
Note that this actually changes the behaviour of rebuilds in certain situations since previously the case would be regenerated in order of transactions, now it will not and so the 'edited' form may overwrite changes made since it was originally received. 

I'm still thinking about the implications of this and may attempt a different approach.

**Comment by snopoke** [comment]
replaced by https://github.com/dimagi/commcare-hq/pull/26644

File: corehq/form_processor/backends/sql/processor.py [hunk 2]
```diff
@@ -210,56 +210,75 @@ def submission_error_form_instance(cls, domain, instance, message):
...
+                else:
+                    normal_form = xform
                 if not (xform.is_deprecated and xform.problem):
                     # don't process deprecated forms which have errors.
                     # see http://manage.dimagi.com/default.asp?243382 for context.
                     # note that we have to use .problem instead of .is_error because applying
                     # the state=DEPRECATED overrides state=ERROR
                     affected_cases.update(case_update.id for case_update in get_case_updates(xform))
 
+            transactions = CaseAccessorSQL.get_case_transactions_for_form(normal_form.form_id, list(affected_cases))
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/form_processor/backends/sql/processor.py]: yes, that is true (the bit out why this works).

**Pull request closed without merging**